tensorflow/compiler/mlir/xla/transforms/tensor_linalg_to_buffer_linalg.cc

```diff
@@ -0,0 +1,152 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file implements logic for transforming Linalg operations with tensor
+// types to memref type and allocate and deallocate buffers using the
+// BufferAssignmentPlacer.
+
+#include "absl/memory/memory.h"
+#include "llvm/ADT/APInt.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"    // TF:llvm-project
+#include "mlir/Dialect/Linalg/IR/LinalgTypes.h"  // TF:llvm-project
+#include "mlir/Dialect/StandardOps/IR/Ops.h"     // TF:llvm-project
+#include "mlir/IR/AffineExpr.h"                  // TF:llvm-project
+#include "mlir/IR/Attributes.h"                  // TF:llvm-project
+#include "mlir/IR/Builders.h"                    // TF:llvm-project
+#include "mlir/IR/Function.h"                    // TF:llvm-project
+#include "mlir/IR/Location.h"                    // TF:llvm-project
+#include "mlir/IR/MLIRContext.h"                 // TF:llvm-project
+#include "mlir/IR/Operation.h"                   // TF:llvm-project
+#include "mlir/IR/PatternMatch.h"                // TF:llvm-project
+#include "mlir/IR/StandardTypes.h"               // TF:llvm-project
+#include "mlir/Pass/Pass.h"                      // TF:llvm-project
+#include "mlir/Transforms/DialectConversion.h"   // TF:llvm-project
+#include "tensorflow/compiler/mlir/xla/ir/lhlo_ops.h"
+#include "tensorflow/compiler/mlir/xla/transforms/buffer_assignment.h"
+#include "tensorflow/compiler/mlir/xla/transforms/rewriters.h"
+
+namespace mlir {
+namespace xla {
+namespace {
+
+class GenericOpConverter
+    : public xla::BufferAssignmentOpConversionPattern<linalg::GenericOp> {
+ public:
+  using xla::BufferAssignmentOpConversionPattern<
+      linalg::GenericOp>::BufferAssignmentOpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      linalg::GenericOp op, ArrayRef<Value> operands,
+      ConversionPatternRewriter& rewriter) const final {
+    auto loc = op.getLoc();
+    SmallVector<Value, 4> args(operands.begin(), operands.end());
+
+    // Update all types to memref types.
+    auto results = op.getOperation()->getResults();
+    for (auto result : results) {
+      auto type = result.getType().cast<ShapedType>();
+      if (!type)
+        op.emitOpError()
+            << "tensor to buffer conversion expects ranked results";
+      auto memrefType = MemRefType::get(type.getShape(), type.getElementType());
+
+      // Compute alloc position and insert a custom allocation node.
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.restoreInsertionPoint(
+          bufferAssignment->computeAllocPosition(result));
+      auto alloc = rewriter.create<AllocOp>(loc, memrefType);
```

Comment:
This will create invalid IR if the memrefType has any unknown dimensions. You need to check for hasStaticShape above and bail out for non-static-shape.

See also https://llvm.discourse.group/t/computing-output-shapes-of-structured-ops-on-tensors/866

Reply:
@silvasean Yes, you are right. This code doesn't take dynamically shaped types into account. We are going to insert an Assert here. Currently, Buffer Assignment doesn't support dynamically shaped types and it's one of our top priorities. We definitely use the link above. Thanks.

Reply:
An assertion would not be a good choice. See https://mlir.llvm.org/getting_started/DeveloperGuide/#assertions-and-crashes-in-passes

You can instead say something like `if (!rankedTensorType.hasStaticShape()) return rewriter.notifyMatchFailure(op, "dynamic shapes not currently supported")`